### DISCUSSION TITLE
How to query touch screen active points?

### DISCUSSION THREAD

**Discussion by adamfierek** [microsoft/microsoft-ui-xaml#8820] [Q&A]
Is it possible using WinUI 3 to check any time if screen is currently touched (and how many touch points are active)? Also what’s the point of PointerPoint IsInContact property since it hold state only from time of event and it not updates when contact disappears. 

I try to check if during Tapped or PointerPressed event, there fingers are touching screen. I tried to register pointer events and store info about touch enter / leave but it’s not precise all time and I don’t believe there is no more simple method to achieve it.

**Comment by DarranRowe** [comment]
Honestly, I don't think this is even possible using the Windows API.
The documentation for [GetPointerInfo](https://learn.microsoft.com/en-us/windows/win32/api/Winuser/nf-winuser-getpointerinfo) implies that the information available is based upon the last pointer message retrieved from the message queue. While this may seem irrelevant, WinUI 3 still runs in a desktop window. This means that there is a Windows API message loop in there. This also means that it will be getting the pointer notifications through the Windows API WM_POINTERxxx messages, such as [WM_POINTERUPDATE](https://learn.microsoft.com/en-us/windows/win32/inputmsg/wm-pointerupdate). WinUI 3 is able to get the pointer ID from this message and wrap it up nicely into a PointerPoint instance.
What's worse is that the documentation for the Windows API [POINTER_INFO](https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-pointer_info) structure implies that you can't even rely on the pointer ID. There is also no function to enumerate currently active pointer IDs. What's worse is that GetPointerInfo also states that you can only obtain pointer info from the thread that it has been posted to. This means that you can't obtain the pointer information from the screen in general.
For PointerPoint.IsInContact, as a hint, pointer input isn't just for touch input. Pen input goes through pointer input, and mouse input can also be configured to go through pointer input. For the mouse, contact is used to state that a mouse button is down. A click is a a pointer down followed by a pointer up. If a pen has proximity detection, then it can be seen as in range and not touching the screen. Contact is then seen as the pen touching the surface of the device.

**Reply by adamfierek** [reply]
Thanks for reply, especially for clarification of IsInContact - now it seems clear for me. What about WinAPI GetPointerFrameInfo()? Is it possible to obtain number of touch contacts using this method?

**Reply by DarranRowe** [reply]
It is possible, but the documentation is problematic. What's more, it still only works with the limitations already mentioned.
The very first sentence in the remarks section states:
"Parallel-mode devices may report pointer input in frames, that is, they may report the state and position of all pointers from that device in a single input report to the system."
The "may" indicates that whether you could get all active touch pointers totally depends on the device itself. If the device doesn't report all of the touch pointers in a single report, then you can't get that information.
What's more, during the event handler for the pointer update event, you can't be sure if you are working with the most recent position or event.
